x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/ApproximateRegExp.java

```diff
@@ -0,0 +1,631 @@
+    public Query toApproximationQuery(StringNormalizer normalizer) throws IllegalArgumentException {
+        Query result = null;
+        switch (kind) {
+            case REGEXP_UNION:
+                result = createUnionQuery(normalizer);
+                break;
+            case REGEXP_CONCATENATION:
+                result = createConcatenationQuery(normalizer);
+                break;
+            case REGEXP_STRING:
+                String normalizedString = normalizer == null ? s : normalizer.normalize(s);
+                result = new TermQuery(new Term("", normalizedString));
+                break;
+            case REGEXP_CHAR:
+                String cs =Character.toString(c);
+                String normalizedChar = normalizer == null ? cs : normalizer.normalize(cs);
+                result = new TermQuery(new Term("", normalizedChar));
+                break;
+            case REGEXP_REPEAT:
+                // Repeat is zero or more times so zero matches = match all
+                result = new MatchAllDocsQuery();
+                break;
+                
+            case REGEXP_REPEAT_MIN:
+            case REGEXP_REPEAT_MINMAX:
+                if (min > 0) {
+                    result = exp1.toApproximationQuery(normalizer);
+                    if(result instanceof TermQuery) {
+                        // Wrap the repeating expression so that it is not concatenated by a parent which concatenates
+                        // plain TermQuery objects together. Boolean queries are interpreted as a black box and not
+                        // concatenated.
+                        BooleanQuery.Builder wrapper = new BooleanQuery.Builder();
+                        wrapper.add(result, Occur.MUST);
+                        result = wrapper.build();
+                    }
+                } else {
+                    // TODO match all was a nice assumption here for optimising .* but breaks 
+                    // for (a){0,3} which isn't a logical match all but empty string or up to 3 a's. 
+//                    result = new MatchAllDocsQuery();
+                    result = new MatchAllButRequireVerificationQuery();
+                }
+                break;
+            case REGEXP_ANYSTRING:
+                // optimisation for .* queries - match all and no verification stage required.
+                result = new MatchAllDocsQuery();
+                break;
+            // All other kinds of expression cannot be represented as a boolean or term query so return an object 
+            // that indicates verification is required
+            case REGEXP_OPTIONAL:
+            case REGEXP_INTERSECTION:
+            case REGEXP_COMPLEMENT:
+            case REGEXP_CHAR_RANGE:
+            case REGEXP_ANYCHAR:
+            case REGEXP_INTERVAL:
+            case REGEXP_EMPTY: 
+            case REGEXP_AUTOMATON:
+                result = new MatchAllButRequireVerificationQuery();
```

Comment:
Can we use a marker rather than a full Lucene query ?

Reply:
See my earlier comment re introducing a base class or marker interface for denoting query clauses that require no verification step. Thoughts?